botbowl/core/pathfinding/cython_pathfinding.pyx

```diff
@@ -152,6 +152,13 @@ cdef class Pathfinder:
     cdef priority_queue[NodePtr] open_set
     cdef mapcpp[double, vector[NodePtr]] risky_sets 
 
+    def __dealloc__(self):
+        cdef NodePtr empty_ptr
+        for i in range(17):
+            for j in range(28):
+                self.locked_nodes[i][j] = empty_ptr
+                self.nodes[i][j] = empty_ptr
+
```

Comment:
This looks oddly specific to the 11 player variant. Does it also work for other variants?

Reply:
It's fine. `locked_nodes` and `nodes` have static size. I choose the size according to the biggest pitch size. If you use a smaller pitch size the out of bounds nodes will just sit there and never be used. I thought this was a better solution than nestling cpp vectors to get a dynamic sizing. 

```python
cdef class Pathfinder:
    .... 
    cdef NodePtr locked_nodes[17][28] # initalized as empty pointers
    cdef NodePtr nodes[17][28] # initalized as empty pointers
```
 
The C equivalent is 
```C
struct {
    ....
    NodePtr    locked_nodes[17][28]; 
    NodePtr    nodes[17][28]; 
} Pathfinder; 
``` 

The cause of the memory leak was that NodePtr's destructor was not called for the nodes in `locked_nodes` and `nodes`, because their type is `NodePtr **` and they don't even know the size of what they're pointing to. And because NodePtr is a `std::shared_ptr` it has the responsibility to free memory on the heap, which was not done. My solution makes sure that the destructor is called.

Reply:
Okay. That's fine for now. Just keep in mind we should be able to support dungeon bowl at some point :)